### ISSUE TITLE
Cache login

### ISSUE THREAD

**Issue by gaynetdinov** [ueberauth/guardian#117]
Hello.

The title of this issue is a bit misleading because I don't know how to describe my question in 2-3 words.

I have a bit weird question. I'm building a mobile web app. My app as any other web app can be added to Home Screen in iOS using Safari. The problem is that if you close such 'standalone' app or just send it to background, the next time you open it you will have to login again because, as I understand, cookies are deleted when you send to background or close 'standalone' web app. Similar issue on StackOverflow http://stackoverflow.com/questions/3813599/iphone-bookmark-to-homescreen-removes-cookies-and-session

So I'd like to cache 'something' on client side using `localStorage` to use this 'something' to be logged in again when I open my 'standalone' web app, so I don't need to submit emai/password every single time when I open my app.
So the question is what should I cache/store on the client side and what should I send to the backend to be logged in?

My current hackity workaround is the following: once a user authenticates using email/password, backend sends Guardian token in the response, client side stores a token in `localStorage`. Then on the login page javascript checks if there is a token in the cache and sends it to an endpoint(i.e. `api/session/exists`) with the following plugs in the pipeline: 
```
plug :fetch_session
plug Guardian.Plug.VerifyHeader
plug Guardian.Plug.LoadResource
```
On the backend side I do the following:
```
case Guardian.Plug.current_resource(conn) do
  nil -> send_resp(conn, 401, "Unauthenticated")
  user -> Guardian.Plug.sign_in(conn, user) |> send_resp(200, "")
end
```

In other words I authenticate a user using API approach(jwt token) to make this user authenticated in browser. This is really weird mixture of API and browser authentication and I believe there is a better approach. Also I'm not sure if such approach is secure. 
Maybe it's possible somehow to communicate a token in the URL(not in header)?

Could you please help me? 

Thank you so much!

**Comment by hassox** [comment]
hey @gaynetdinov 

Local storage definitely seems like the thing you want to use here. I think there's a couple of ways I'd handle this. 

1. Make the client use Header based (API style) authentication always. This is great if you're using a single page app and all your requests are done over ajax. Not so good if you have normal server generated pages with straight up HTTP
2. Provide your client with a "re-login" method, which it looks like you have. This would look like
    
    1. open the app
    2. If there's a token in local storage
    3. Shoot a request using the token in the header to the 'relogin' endpoint
    4. The relogin endpoint would log the user into the session
    5. In your router pipeline allow both session and header verification

The challenge with the 2nd is you have to do it on each page re-fresh. How do you know if you're already logged into the session. 

I'd tend to lean towards the first option because it minimizes the weirdness. Option 2 is had to keep in sync and you'll need a bunch of extra stuff to try and keep on top of it.

**Comment by gaynetdinov** [comment]
I think I'd go with second option. All my requests are done via ajax, so once I get a response from the 'relogin' endpoint, the client side has a proper session and all ajax requests work just fine. 

Thank you again, @hassox!
